betty/cache/__init__.py

```diff
@@ -0,0 +1,112 @@
+"""
+Provide the Cache API.
+"""
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from pathlib import Path
+from typing import Sequence, Self, Generic, TypeAlias, AsyncContextManager, Any
+
+from typing_extensions import TypeVar, deprecated
+
+from betty.concurrent import _AsyncLock
+
+
+class _HasScopes:
+    @property
+    def scopes(self) -> Sequence[str]:
+        raise NotImplementedError
+
+
+CacheItemValueT = TypeVar('CacheItemValueT')
+CacheItemValueCoT = TypeVar('CacheItemValueCoT', covariant=True)
+CacheItemValueContraT = TypeVar('CacheItemValueContraT', contravariant=True)
+
+
+class CacheItem(_HasScopes, Generic[CacheItemValueCoT]):
+    @property
+    def id(self) -> str:
+        """
+        Get this cache item's ID.
+        """
+        raise NotImplementedError
+
+    async def value(self) -> CacheItemValueCoT:
+        """
+        Get this cache item's value.
+        """
+        raise NotImplementedError
+
+    @property
+    def modified(self) -> int | float:
+        """
+        Get the time this cache item was last modified, in seconds.
+        """
+        raise NotImplementedError
+
+
+CacheItemT = TypeVar('CacheItemT', bound=CacheItem[Any])
+CacheItemCoT = TypeVar('CacheItemCoT', covariant=True)
+
+
+CacheItemValueSetter: TypeAlias = Callable[[CacheItemValueT], Awaitable[None]]
+
+
+class Cache(Generic[CacheItemCoT, CacheItemValueContraT], _HasScopes):
```

Comment:
Can we simplify this by being generic only over the value type?

Reply:
This exists only for the binary file cache but it's proven not to be all that useful to get a cache item's path from the item (when there is no item).